app/models/contract.rb

```diff
@@ -63,9 +63,14 @@ def self.issued_by(issuer, *ids)
   end
 
   scope :permitted_for, ->(user) {
-    next all unless user.forbidden_some_services?
-
-    where(service_id: user.member_permission_service_ids)
+    case user.permitted_services_status
+    when :none
+      none
+    when :all
+      merge(provided_by(user.account))
```

Comment:
@akostadinov 

> Maybe use scope `provider_account: user.provider_account` although if we don't have these fields and I think we don't, then it may be complicated. So for now it is not worse than before at least.

> For clarity, I would suggest to refactoring it into something like:
> 
> ```ruby
> case permitted_services_status
> when :all, :none
>   public_send permitted_services_status
> when :selected
>   where(service_id: user.member_permission_service_ids)
> else
>   {} # but what else could be the value? doesn't `{}` in this case mean the same as `all`?
> end
> ```

As I was touching this scope anyway, I decided to fix my own `FIXME`...

Basically, when the permission is `:all` I am just applying the `provided_by(user.account)` scope, which should return all the contracts that belong to the provider account. I verified in a test that without this, the scope can return cinstances that belong to a different provider - **only** if this `permitted_for` scope is used separately, which currently is not the case - it's allways appended to `provided_cinstances`. However, in case a developer makes a mistake and uses this scope separately, this will prevent from tenant leak.

Now, it also means that currently in the code whenever `account.provided_cinstances.permitted_for(user)` is used, the same scope `provided_by` will be merged 2 times. However, I think that ActiveRecord resolves this correctly, and the SQL query is not becoming more complex.

But still not sure whether it's a good idea. WDYT?

Reply:
Super weird... retesting worked https://app.circleci.com/pipelines/github/3scale/porta/30235/workflows/2b5cd8d5-747e-403f-917a-f5d9ac278266

Also, I saw the same locally :flushed: 

Some test runs failed, but then they started to succeed... super weird.